File: http-streams/src/main/java/http/Basics.java
package http;

import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.Http;
import akka.http.javadsl.IncomingConnection;
import akka.http.javadsl.ServerBinding;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;

import static java.lang.System.out;

import java.util.concurrent.CompletionStage;

public class Basics {
  public static void main(String[] args) {
    new Basics().run();
  }
  public void run() {
    //#init
    ActorSystem system = ActorSystem.create();
    Http http = Http.get(system);
    //#init

    //#bind
    Source<IncomingConnection, CompletionStage<ServerBinding>> bind =
      http.bind(ConnectHttp.toHost("localhost", 0));
    //#bind

    //#run
    Materializer materializer = ActorMaterializer.create(system);

    CompletionStage<ServerBinding> bound = bind
      .to(Sink.foreach(c -> out.println("Got HTTP connection!")))
      .run(materializer);
    //#run

    //#log
    bound.thenAccept(b ->
      out.println("Bound on " + b.localAddress()));
    //#log
  }

  void flow() {
    IncomingConnection connection = null;
    //#flow
    Flow<HttpResponse, HttpRequest, NotUsed> flow =
      connection.flow();

    // Materialize and run it yourself
    //#flow
  }
  void handleWith() {
    IncomingConnection connection = null;
    Materializer materializer = null;
    Flow<HttpRequest, HttpResponse, NotUsed> myflow = null;
    //#handleWith
    // Construct your flow:
    Flow<HttpRequest, HttpResponse, NotUsed> flow = myflow;

    // Use it to handle the connection:
    connection.handleWith(flow, materializer);
    //#handleWith
  }
  void bindAndHandle() {
    Materializer materializer = null;
    Flow<HttpRequest, HttpResponse, NotUsed> myflow = null;
    Http http = null;
    //#bindAndHandle
    // Construct your flow:
    Flow<HttpRequest, HttpResponse, NotUsed> flow = myflow;

    // Use it to handle connections:
    http.bindAndHandle(
      flow,
      ConnectHttp.toHost("localhost", 8080),
      materializer);
    //#bindAndHandle
  }
}

File: http-streams/src/main/scala/http/Basics.scala
package sc.http

import scala.concurrent.Future

import akka.NotUsed
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.Http.IncomingConnection
import akka.http.scaladsl.Http.ServerBinding
import akka.http.scaladsl.model.HttpRequest
import akka.http.scaladsl.model.HttpResponse
import akka.stream.ActorMaterializer
import akka.stream.Materializer
import akka.stream.scaladsl.Flow
import akka.stream.scaladsl.Sink
import akka.stream.scaladsl.Source

class Basics {
  def run() = {
    //#init
    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    val http = Http()
    //#init

    //#bind
    val binding: Source[IncomingConnection, Future[ServerBinding]] =
      http.bind("localhost", port = 0)
    //#bind

    //#run

    val bound: Future[ServerBinding] = binding
      .to(Sink.foreach(c => println(s"Got HTTP connection!")))
      .run()
    //#run

    implicit val ec = materializer.executionContext
    //#log
    bound.foreach(b =>
      println("Bound on " + b.localAddress))
    //#log
  }

  def flow() = {
    val connection: IncomingConnection = null
    //#flow
    val flow: Flow[HttpResponse, HttpRequest, NotUsed] =
      connection.flow

    // Materialize and run it yourself
    //#flow
  }
  def handleWith() = {
    val connection: IncomingConnection = null
    implicit val materializer: Materializer  = null
    val myflow: Flow[HttpRequest, HttpResponse, NotUsed] = null
    //#handleWith
    // Construct your flow:
    val flow: Flow[HttpRequest, HttpResponse, NotUsed] = myflow

    // Use it to handle the connection:
    connection.handleWith(flow)
    //#handleWith
  }
  def bindAndHandle() {
    implicit val system = ActorSystem()
    implicit val materializer: Materializer = null
    val myflow: Flow[HttpRequest, HttpResponse, NotUsed] = null
    val http = Http()
    //#bindAndHandle
    // Construct your flow:
    val flow: Flow[HttpRequest, HttpResponse, NotUsed] = myflow

    // Use it to handle connections:
    http.bindAndHandle(
      flow,
      "localhost",
      port = 8080)
    //#bindAndHandle
  }
}
